feat: Implement Level 2 card processing support (5686)

Narek13 opened the pull request:
## Level 2 Card Processing Implementation

Implements Level 2 card processing for PayPal ACDC (Advanced Credit and Debit Card) payments, enabling lower interchange rates for eligible B2B transactions on Visa and Mastercard.

### Changes

**New Classes:**
- `PaymentLevelEligibility` - Validates payment eligibility for Level 2/3 processing based on merchant country, shop currency, payment method, and feature setting
- `PaymentLevelHelper` - Builds Level 2 supplementary card data structure from Amount entities

**Modified Classes:**
- `PurchaseUnit` - Added `supplementary_data` constructor parameter and array output support
- `PurchaseUnitFactory` - Integrated Level 2 data building in both `from_wc_cart()` (checkout) and `from_wc_order()` (pay-now)

**Settings:**
- Added `payment_level_processing` feature setting to enable/disable Level 2/3 processing

<img width="1048" height="469" alt="Screenshot 2025-12-26 at 17 37 03" src="https://github.com/user-attachments/assets/46f38863-e099-4696-8b8b-e37ab2cfbc13" />

### How It Works

Level 2 data is automatically included for eligible transactions when:
1. Feature setting `payment_level_processing` is enabled
2. Merchant country is US
3. Shop currency is USD
4. Payment method is ACDC (credit card gateway)

**Supports both contexts:**
- **Checkout flow** - Level 2 data built from cart Amount before WC order exists
- **Pay-now flow** - Level 2 data built from existing WC order Amount

**Level 2 Data Structure:**
```json
{
  "supplementary_data": {
    "card": {
      "level_2": {
        "invoice_id": "INV_6789ABCD",
        "tax_total": {
          "currency_code": "USD",
          "value": "8.50"
        }
      }
    }
  }
}
```

### Card Network Validation

Card network validation (Visa/Mastercard only) is **not** implemented at order creation time because the card brand is not available until after the customer enters card details in PayPal's hosted fields. PayPal automatically filters Level 2 data for non-Visa/Mastercard transactions. See [PCP-5686 comment](https://inpsyde.atlassian.net/browse/PCP-5686?focusedCommentId=365263) for detailed analysis and implementation options.

### Filters

**`woocommerce_paypal_payments_level2_invoice_id`**
- Customize the Level 2 invoice ID (default: generated unique ID with `INV_` prefix, max 127 chars)

**`woocommerce_paypal_payments_level_processing_eligible`**
- Override eligibility determination for specific payments

**`woocommerce_paypal_payments_level_processing_countries`**
- Modify allowed merchant countries (default: `['US']`)

**`woocommerce_paypal_payments_level_processing_currencies`**
- Modify allowed currencies (default: `['USD']`)

**`woocommerce_paypal_payments_level_processing_payment_methods`**
- Modify allowed payment methods (default: `['ppcp-credit-card-gateway']`)

### Testing Notes

- [ ] Enable `payment_level_processing` feature flag
- [ ] Test with US merchant, USD currency, ACDC payment in **checkout**
- [ ] Test with existing order in **pay-now** context
- [ ] Verify Level 2 data in PayPal Orders API request payload
- [ ] Test with ineligible conditions (non-US, non-USD, non-ACDC)
- [ ] Verify no Level 2 data sent when feature flag is disabled
- [ ] Test invoice_id filter customization

### Future Work

- Level 3 implementation with line-item details (separate ticket)

Narek13 committed:
feat: Add PaymentLevelEligibility helper for Level 2/3 processing validation

Introduces PaymentLevelEligibility class to determine if an order qualifies
for Level 2/Level 3 card processing based on:
- flag enablement (payment_level_processing setting)
- Merchant country (US only by default)
- Order currency (USD only by default)
- Payment method (ACDC)

The class provides filterable eligibility criteria allowing merchants and
developers to customize Level 2/3 processing requirements. Card network
validation (Visa/Mastercard) is intentionally omitted as card brand is
not available at order creation time.

Narek13 committed:
feat: Add `PaymentLevelHelper` for building Level 2 card processing data

Introduces `PaymentLevelHelper` class to build Level 2 supplementary card data
for PayPal Orders API. Maps WooCommerce order information to PayPal's Level 2
processing structure:
- customer_reference: Order ID (max 17 characters, filterable)
- tax_total: Order tax amount as PayPal Money object

Level 2 data enables lower interchange rates for eligible B2B transactions
on Visa/Mastercard. The helper is designed for extensibility with Level 3
support (line-item details) planned for future implementation.

Narek13 committed:
feat: Add `supplementary_data` support to PurchaseUnit for Level 2/3 processing

Extends PurchaseUnit entity and factory to support Level 2/3 card processing:

PurchaseUnit:
- Adds supplementary_data property for storing Level 2/3 card data
- Includes `supplementary_data` in to_array() output when present

PurchaseUnitFactory:
- Injects PaymentLevelHelper and PaymentLevelEligibility dependencies
- Builds Level 2 data in `from_wc_order()`` for eligible ACDC transactions
- Validates eligibility (US merchants, USD currency, credit card payment)

Narek13 committed:
feat: Add `payment_level_processing` setting

Introduces `payment_level_processing` boolean setting to control Level 2/3
card processing functionality. When enabled, supplementary card data is
included in PayPal Orders API requests for eligible ACDC transactions.

Narek13 committed:
feat: Add `payment_level_processing` setting save/get logic

Narek13 committed:
feat: Add `payment_level_processing` setting JS logic

Narek13 committed:
feat: Add all appropriate services

Narek13 committed:
chore: Add/fix unit tests

Narek13 committed:
chore: fix psalm

Narek13 committed:
refactor: Adapt PaymentLevelEligibility for cart-based eligibility checks

Refactors PaymentLevelEligibility to support cart-based contexts where
WC_Order doesn't exist yet:

- Accepts payment_method string parameter instead of WC_Order
- Uses shop-level currency via CurrencyGetter instead of order currency
- Uses merchant country from shop settings instead of order data
- Enables eligibility checks during checkout before order creation

This allows Level 2 processing eligibility to be determined when creating
PayPal orders from cart during checkout, not just for existing WC orders
in pay-now context.

Narek13 committed:
refactor: Adapt PaymentLevelHelper for cart-based Level 2 data generation

Refactors PaymentLevelHelper to support cart-based contexts where
WC_Order doesn't exist yet:

- Accepts Amount entity instead of WC_Order
- Uses invoice_id field instead of customer_reference (per PayPal API spec)
- Generates unique invoice_id with INV_ prefix via uniqid()
- Extracts tax_total from Amount breakdown structure
- Supports filterable invoice_id via woocommerce_paypal_payments_level2_invoice_id

This allows Level 2 supplementary card data to be built when creating
PayPal orders from cart during checkout, not just for existing WC orders
in pay-now context. The invoice_id field matches PayPal's documented
Level 2 structure with 127 character max length.

Narek13 committed:
feat: Integrate Level 2 processing in PurchaseUnitFactory

Adds Level 2 card processing support to both cart and order-based
purchase unit creation:

from_wc_cart():
- Accepts payment_method parameter from CreateOrderEndpoint
- Checks eligibility and builds Level 2 data for checkout context
- Extracts supplementary_data from PaymentLevelHelper result

from_wc_order():
- Checks eligibility using order payment method
- Builds Level 2 data for pay-now context
- Extracts supplementary_data from PaymentLevelHelper result

Both methods now consistently extract and pass supplementary_data
from PaymentLevelHelper output to PurchaseUnit constructor, enabling
reduced interchange rates on eligible ACDC transactions in both
checkout and pay-now flows.

Narek13 committed:
test: Update PurchaseUnitFactoryTest for Level 2 processing integration

Updates test suite to reflect PurchaseUnitFactory changes:

Narek13 committed:
chore: Update appropriate services

Narek13 committed:
chore: Update PurchaseUnitFactory test

Narek13 committed:
chore: Fix Psalm

Dinamiko commented on modules/ppcp-api-client/src/Helper/PaymentLevelEligibility.php:
```diff
@@ -0,0 +1,102 @@
+
+namespace WooCommerce\PayPalCommerce\ApiClient\Helper;
+
+use WC_Order;
```
`WC_Order` seems not used.

Dinamiko commented on modules/ppcp-api-client/src/Helper/PaymentLevelHelper.php:
```diff
@@ -0,0 +1,100 @@
+
+namespace WooCommerce\PayPalCommerce\ApiClient\Helper;
+
+use WC_Order;
```
`WC_Order` seems not used.

Dinamiko reviewed (approved):
PR looks good, I've approved it. 

Perhaps I'm missing some unit tests for `PaymentLevelEligibility` class which  is critical as it determines whether Level 2 data is sent to PayPal. I would suggest unit tests covering, e.g. setting disabled, invalid country, currency, payment method...

Narek13 commented:
@Dinamiko Thanks

> Perhaps I'm missing some unit tests for PaymentLevelEligibility class which is critical as it determines whether Level 2 data is sent to PayPal. I would suggest unit tests covering, e.g. setting disabled, invalid country, currency, payment method...

I have added some tests inside `PurchaseUnitFactoryTest.php` which makes sure the `build` is not called when is not eligible.

Ps. Also added the `PaymentLevelEligibilityTest`

Narek13 committed:
chore: remove unused namespace

Narek13 committed:
test: add PaymentLevelEligibility unit tests
